src/language/patterns.md

```diff
@@ -0,0 +1,334 @@
+a constant returns `true`.
+  ```dart
+  switch (code) {
+    case >= first && <= last: ...    
+    // Matches if the value of code is between the constant values of first and last
+  }
+  ```  
+- [**Null-check**][check] or [**null-assert**][assert] **patterns**: the match is first on whether the object is not null, then on the value.
+  ```dart
+  switch (maybe) {
+    case var s?: ... 
+    // Matches if maybe is not null, and then against the inner pattern "var s"
+  }
+  ```  
+- [**Object patterns**][object]: the match exposes getters on the object type.
+  ```dart
+  switch (Shape shape) {
+    case Rect(width: var w, height: var h): ...
+    // Matches if shape is of type Rect, and then against the properties of Rect
+  }
+  ```  
+
+Values destructured by a pattern in a case become local variables,
+only available in the body of that case. 
+
+### Object destructuring
+
+[Object patterns][object] match against named object types, allowing
+you to destructure their data using the getters the object's class already exposes.
+
+For example, you can use object destructuring in a [for-in loop][for] to iterate-over
+and destructure the [`MapEntry`][] objects that a `<Map>.entries` call
+returns:
+
+```dart
+Map<String, int> hist = // ...
+
+for (var MapEntry(key: key, value: count) in hist.entries) {
+  print('$key occurred $count times');
+}
+```
+
+The object pattern checks that `hist.entries` has the named type `MapEntry`,
+and then recurses into the named field subpatterns `key` and `value`.
+It calls the `key` getter and `value` getter on the `MapEntry` in each iteration,
+and binds the results to local variables `key` and `count`, respectively.
+
+Binding the result of a getter call to a variable of the same name is a common
+use case, so object patterns can also infer the getter name from the
+[variable subpattern][variable]. This allows you to simplify the variable pattern
+from something redundant like `key: key` to just `:key`:
+
+```dart
+for (var MapEntry(:key, value: count) in hist.entries) { 
+  print('$key occurred $count times');
+}
+```
+
+### Algebraic datatypes 
+
+Object destructuring and switch cases are conducive to writing
+code in an [algebraic datatype][] style: 
+when you have a family of related types and an operation that needs
+specific behavior for each type.
+
+Instead of implementing the operation as an instance method for every respective type,
+you can keep the operations in a single function, switching over subtypes for
+variations in behavior:
+
+```dart
+abstract class Shape {
+  double calculateArea();
+}
+
+class Square implements Shape {
+  final double length;
+  Square(this.length);
+
+  double calculateArea() => length * length;
+}
+
+class Circle implements Shape {
+  final double radius;
+  Circle(this.radius);
+
+  double calculateArea() => math.pi * radius * radius;
+}
+
+double calculateArea(Shape shape) =>
```

Comment:
This code is an error as written, since switch expressions must be exhaustive.  You could make `Shape` be `sealed`, or you could use a switch statement, or you could add a default case.  Given that you are talking about algebraic datatypes here, I think the first solution is the correct one, since that's the use case that `sealed` addresses.